pages/_app.js

```diff
@@ -63,29 +64,65 @@ export default function App({ Component, pageProps }) {
   if (error) return <div>failed to load</div>;
   if (isLoading) return <div>loading...</div>;
 
-  function handleNewPet(newPet) {
-    newPet.id = uid();
-    const petsWithNewPet = [newPet, ...pets];
-    setPets(petsWithNewPet);
-  }
+  // add new pet
+  async function handleNewPet(newPet) {
+    try {
+      const response = await fetch(`/api/pets`, {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify(newPet),
+      });
 
-  function handleUpdate(updatedPet) {
-    const updatedPets = pets.map((pet) => {
-      if (updatedPet.id !== pet.id) {
-        return pet;
+      if (response.ok) {
+        mutate();
+      } else {
+        console.error("Failed to create pet");
       }
-      return updatedPet;
-    });
-    setPets(updatedPets);
+    } catch (error) {
+      console.error("Error creating pet:", error);
+    }
   }
 
-  function handleDelete(petToDelete) {
-    const petsWithoutDeletedPet = pets.filter(
-      (pet) => pet.slug !== petToDelete.slug
-    );
+  // update existing pet
+  async function handleUpdate(updatedPet) {
+    try {
+      const response = await fetch(`/api/pets/${updatedPet._id}`, {
+        method: "PUT",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: JSON.stringify(updatedPet),
+      });
+
+      if (response.ok) {
+        mutate();
+        console.log("success");
+      } else {
+        console.error(`Failed to update ${pet.petName}:`);
+      }
+    } catch (error) {
+      console.error(`Error updating ${pet.petName}:`, error);
+    }
+  }
 
-    router.push("/");
-    setPets(petsWithoutDeletedPet);
+  // delete existing pet
+  async function handleDelete(petToDelete) {
+    try {
+      const response = await fetch(`/api/pets/${petToDelete._id}`, {
+        method: "DELETE",
+      });
+      if (response.ok) {
+        mutate();
+        router.push("/");
```

Comment:
You may not need the `mutate` here since you're changing the route immediately after that. Changing the route triggers a revalidation already. I'm not too sure, though. You can try removing the `mutate` and see if it still works.